src/Common/ConcurrentBoundedQueue.h

```diff
@@ -138,11 +138,29 @@ class ConcurrentBoundedQueue
     }
 
     /// Returns false if queue is (finished and empty) or (object was not popped during timeout)
-    [[nodiscard]] bool tryPop(T & x, UInt64 milliseconds = 0)
+    [[nodiscard]] bool tryPop(T & x, UInt64 milliseconds)
     {
         return popImpl(x, milliseconds);
     }
 
+    /// Returns false if queue is empty.
+    [[nodiscard]] bool tryPop(T & x)
+    {
+        // we don't use popImpl to avoid CV wait
+        {
+            std::lock_guard queue_lock(queue_mutex);
+
+            if (queue.empty())
```

Comment:
`popImpl()` checks `is_finished`, should it be done here as well?

Reply:
`is_finished` is checked because `popImpl` waits on CV. We don't have a need for it because the `tryPop` without timeout will just check the queue under lock and instantly return. 
Notice also how `popImpl` will still return results even when the `is_finished` is set to `true` so we have the same behaviour but without using CV.